compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/body/resolve/FirExpressionsResolveTransformer.kt

```diff
@@ -374,64 +377,90 @@ open class FirExpressionsResolveTransformer(transformer: FirBodyResolveTransform
+                    source = assignmentOperatorStatement.source
+                    rValue = resolvedOperatorCall
+                    calleeReference = when {
+                        lhsIsVar -> lhsReference!!
+                        else -> buildErrorNamedReference {
+                            source = when (leftArgument) {
+                                is FirFunctionCall -> leftArgument.source
+                                is FirQualifiedAccess ->
+                                    leftArgument.calleeReference.source
+                                else -> leftArgument.source
                             }
-                        }
-                        (leftArgument as? FirQualifiedAccess)?.let {
-                            dispatchReceiver = it.dispatchReceiver
-                            extensionReceiver = it.extensionReceiver
+                            diagnostic = if (lhsSymbol == null) ConeVariableExpectedError() else ConeValReassignmentError(lhsSymbol)
                         }
                     }
-                assignment.transform(transformer, ResolutionMode.ContextIndependent)
-            }
-            else -> {
-                val operatorCallSymbol = operatorCallReference?.candidateSymbol
-                val assignmentCallSymbol = assignCallReference?.candidateSymbol
+                    (leftArgument as? FirQualifiedAccess)?.let {
+                        dispatchReceiver = it.dispatchReceiver
+                        extensionReceiver = it.extensionReceiver
+                    }
+                }
+            return assignment.transform(transformer, ResolutionMode.ContextIndependent)
+        }
 
-                requireNotNull(operatorCallSymbol)
-                requireNotNull(assignmentCallSymbol)
+        fun reportAmbiguity(): CompositeTransformResult<FirStatement> {
+            val operatorCallSymbol = operatorCallReference?.candidateSymbol
+            val assignmentCallSymbol = assignCallReference?.candidateSymbol
 
-                buildErrorExpression {
-                    source = assignmentOperatorStatement.source
-                    diagnostic = ConeOperatorAmbiguityError(listOf(operatorCallSymbol, assignmentCallSymbol))
-                }.compose()
-            }
+            requireNotNull(operatorCallSymbol)
+            requireNotNull(assignmentCallSymbol)
+
+            return buildErrorExpression {
+                source = assignmentOperatorStatement.source
+                diagnostic = ConeOperatorAmbiguityError(listOf(operatorCallSymbol, assignmentCallSymbol))
+            }.compose()
+        }
+
+        return when {
+            assignIsResolvable && !lhsIsVar -> chooseAssign()
+            assignIsResolvable && !operatorIsResolvableAndReturnTypeMatches -> chooseAssign()
+            !assignIsResolvable && operatorIsResolvable -> chooseOperator()
+            !assignIsResolvable && !operatorIsResolvable -> chooseAssign()
+            else -> reportAmbiguity()
```

Comment:
These simplifications look great! But I'm not sure I followed them correctly. :\

Previously, assign was chosen when `operatorIsError || (!lhsIsVar && !assignIsError)`. Its rewritten form would be: `!operatorIsResolvable || (!lhsIsVar && assignIsResolvable)`, right?

Once the former is false, i.e., `!operatorIsError && (lhsIsVar || assignIsError)`, then `assignIsError` is checked, whose counterpart would be: `!assignIsResolvable`.

------

Okay... so, from `!operatorIsResolvable || (!lhsIsVar && assignIsResolvable)`, the first line is the rhs of the logic: `assignIsResolvable && !lhsIsVar`. Then, where is lhs of the logic? Oh, maybe 3rd line, `!assignIsResolvable && !operatorIsResolvable` covers part of it? Speaking of it, that line is read...weird: if neither assign nor operator is resolvable, why should we choose assign? :p

Another thing that I can'f quite follow is how `operatorIsResolvableAndReturnTypeMatches` is added (in terms of previous logics).

Reply:
OK, let me rename `operatorIsResolvableAndReturnTypeMatch` to simply `operatorReturnTypeMatches`, where it's implicit that if operator is not resolvable, then the return type doesn't match. Next in the when branches, I will make the cases more redundant, but hopefully easier to follow.

To your last question, this PR basically introduces a new flag `operatorReturnTypeMatches` and make decisions based on this new piece of information.

Reply:
I have split the commit so now commit#1 is pure refactoring with no logic change.

Reply:
Thanks for the split! That made reviewing much easier!

The only thing that still bothered me is:

> `!assignIsResolvable && !operatorIsResolvable` ... Speaking of it, ... if neither assign nor operator is resolvable, why should we choose assign?

Reply:
I am not sure why the original FIR implementation did this (https://github.com/JetBrains/kotlin/blob/d05ef3a63589b6362702ef95d318e7680df3d026/compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/body/resolve/FirExpressionsResolveTransformer.kt#L392). It seems to be arbitrary. Personally I think picking either one is fine since the reported error message should be informative either way. FE1.0 seems to be picking "operator" instead when non is resolved, though. @mglukhikh is it intended that in FIR we pick "assign" instead of "operator"?

Reply:
I would also add a comment in the code that neither is resolvable, and we just pick one of them anyway.